From 213c9705b64d66ad5a8ebcc1943817cddcf2f60c Mon Sep 17 00:00:00 2001
From: saravananm <saravananm@ami.com>
Date: Tue, 17 Oct 2023 15:11:20 +0530
Subject: [PATCH] fix-for-get-sdr-count

Description :
 IPMI SDR count raw command shows wrong sdr count.

Tested:

Build Successfully and Verified,its Working As Expected.

Before Fix:

root@AMIOT-c2045e432c9b:~# ipmitool raw  0x04 0x20 0x01
 46 81 5f 32 2e 65
root@AMIOT-c2045e432c9b:~# ipmitool sdr info
SDR Version                         : 0x51
Record Count                        : 153
Free Space                          : unspecified
Most recent Addition                : NA

root@AMIOT-c2045e432c9b:~# ipmitool raw  0x04 0x20 0x01
 9a 81 6c 32 2e 65
root@AMIOT-c2045e432c9b:~# ipmitool sdr info
SDR Version                         : 0x51
Record Count                        : 154
Free Space                          : unspecified

Working as excepted.

Signed-off-by: saravananm <saravananm@ami.com>
---
 src/sensorcommands.cpp | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/src/sensorcommands.cpp b/src/sensorcommands.cpp
index 6c9d678..314af41 100644
--- a/src/sensorcommands.cpp
+++ b/src/sensorcommands.cpp
@@ -1384,6 +1384,7 @@ static int
     size_t lastRecord = getNumberOfSensors() + fruCount +
                         ipmi::storage::type12Count +
                         ipmi::storage::nmDiscoverySDRCount - 1;
+    recordData.clear();
     if (recordID == lastRecordIndex)
     {
         recordID = lastRecord;
@@ -1797,6 +1798,20 @@ static ipmi::RspType<uint8_t, // respcount
                     sdrCount++;
                 }
             }
+            else if (hdr->record_type == get_sdr::SENSOR_DATA_COMPACT_RECORD)
+            {
+                get_sdr::SensorDataCompactRecord* recordData =
+                    reinterpret_cast<get_sdr::SensorDataCompactRecord*>(
+                        record.data());
+                if (ctx->lun == recordData->key.owner_lun)
+                {
+                    sdrCount++;
+                }
+            }
+            else if (hdr->record_type == get_sdr::SENSOR_DATA_FRU_RECORD)
+            {
+                sdrCount++;
+            }
         }
         sdrCount +=
             ipmi::storage::type12Count + ipmi::storage::nmDiscoverySDRCount;
-- 
2.25.1

